lgc/patch/Continufy.cpp

```diff
@@ -0,0 +1,205 @@
+  switch (rtStage) {
+  case RtStage::ShaderStageRayGeneration:
+    return CpsLevel::RayGen;
+  case RtStage::ShaderStageClosestHit:
+  case RtStage::ShaderStageMiss:
+  case RtStage::ShaderStageCallable:
+    return CpsLevel::ClosestHit_Miss_Callable;
+  case RtStage::ShaderStageAnyHit:
+    return CpsLevel::AnyHit_CombinedIntersection_AnyHit;
+  case RtStage::ShaderStageIntersection:
+    return CpsLevel::Intersection;
+  default:
+    llvm_unreachable("Unknown raytracing shader type.");
+  }
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The analyses that are still valid after this pass)
+PreservedAnalyses Continufy::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  LLVM_DEBUG(dbgs() << "Run the Continufy pass \n");
+  m_pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
+  LLVMContext &context = module.getContext();
+
+  llvm_dialects::Builder builder(context);
+  SmallVector<Instruction *> tobeErased;
+
+  for (auto &fn : make_early_inc_range(module.functions())) {
+    MDNode *continufyStage = fn.getMetadata("continufy.stage");
+    Function *fnPtr = &fn;
+    std::optional<int32_t> currentRtStage;
+    if (continufyStage) {
+      fnPtr = insertCpsArguments(fn);
+      currentRtStage = mdconst::extract<ConstantInt>(continufyStage->getOperand(0))->getSExtValue();
+      CpsLevel level = getLevelOfRtStage(currentRtStage.value());
+      setCpsFunctionLevel(fn, level);
+    }
+
+    // Translate call instruction with %continufy.stage into lgc.cps.await() with continuation reference.
+    for (auto &block : *fnPtr) {
+      for (auto &inst : block) {
+        if (!isa<CallInst>(inst))
+          continue;
+        auto *calleeStage = inst.getMetadata("continufy.stage");
+        if (!calleeStage)
+          continue;
+
+        auto &call = cast<CallInst>(inst);
+        assert(call.getCallingConv() == CallingConv::SPIR_FUNC);
+        auto *called = call.getCalledOperand();
+
+        builder.SetInsertPoint(&call);
+        auto *continuationRef = builder.CreatePtrToInt(called, IntegerType::get(context, 32));
+        CpsLevel calleeLevel =
+            getLevelOfRtStage(mdconst::extract<ConstantInt>(calleeStage->getOperand(0))->getSExtValue());
+        // RayGen level is zero, so it does not need a logic OR here.
+        if (calleeLevel != CpsLevel::RayGen)
+          continuationRef = builder.CreateOr(continuationRef, builder.getInt32((uint32_t)calleeLevel));
+
+        SmallVector<Value *> callArgs(call.args());
+        auto *newCall = builder.create<AwaitOp>(call.getType(), continuationRef, 1u << (unsigned)calleeLevel, callArgs);
```

Comment:
Why not let `getLevelOfRtStage` compute the shifted level?

Reply:
I don't see why we want to let `getLevelOfRtStage` calculate shifted level. We only need to shift the level when we want to generate the mask. We also need the un-shifted level.